pyro/distributions/empirical.py

```diff
@@ -0,0 +1,160 @@
+from pyro.distributions.util import log_sum_exp
+
+
+@copy_docs_from(TorchDistribution)
+class Empirical(TorchDistribution):
+    r"""
+    Empirical distribution associated with the sampled data.
+    """
+
+    arg_constraints = {}
+
+    def __init__(self):
+        self._samples = None
+        self._log_weights = None
+        self._categorical = None
+        self._samples_buffer = []
+        self._weights_buffer = []
+        super(TorchDistribution, self).__init__()
+
+    def _append_from_buffer(self, tensor, buffer):
+        """
+        Append values from the buffer to the finalized tensor, along the
+        leftmost dimension.
+
+        :param torch.Tensor tensor: tensor containing existing values.
+        :param list buffer: list of new values.
+        :return: tensor with new values appended at the bottom.
+        """
+        buffer_tensor = torch.stack(buffer, dim=0)
+        return torch.cat([tensor, buffer_tensor], dim=0)
+
+    def _finalize(self):
+        """
+        Appends values collected in the samples/weights buffers to their
+        corresponding tensors.
+        """
+        if not self._samples_buffer:
+            return
+        self._samples = self._append_from_buffer(self._samples, self._samples_buffer)
+        self._log_weights = self._append_from_buffer(self._log_weights, self._weights_buffer)
+        self._categorical = Categorical(logits=self._log_weights)
+        # Reset buffers.
+        self._samples_buffer, self._weights_buffer = [], []
+
+    @property
+    def sample_size(self):
+        """
+        Number of samples that constitute the empirical distribution.
+
+        :return int: number of samples collected.
+        """
+        self._finalize()
+        if self._samples is None:
+            return 0
+        return self._samples.size(0)
+
+    def add(self, value, log_weight=None, weight=None):
+        """
+        Adds a new data point to the sample. The values in successive calls to
+        ``add`` must have the same tensor shape and size. Optionally, an
+        importance weight can be specified via ``log_weight`` or ``weight``
+        (default value of `1` is used if not specified).
+
+        :param torch.Tensor value: tensor to add to the sample.
+        :param torch.Tensor weight: log weight (optional) corresponding
+            to the sample.
+        :param torch.Tensor log_weight: weight (optional) corresponding
+            to the sample.
+        """
+        if is_validation_enabled():
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = torch.Tensor if value.dtype in (torch.int32, torch.int64) else value.type()
```

Comment:
This doesn't work with cuda int tensors. Consider using something that preserves CUDA and device, e.g.
```py
weight_type = value.new_empty(1).float().type() if ... else value.type()
```

Reply:
Yes, let me change that.